libdrgn/memory_reader.c

```diff
@@ -273,19 +273,20 @@ struct drgn_error *drgn_read_memory_file(void *buf, uint64_t address,
 					 void *arg, bool physical)
 {
 	struct drgn_memory_file_segment *file_segment = arg;
+	uint64_t file_offset = file_segment->file_offset + offset;
+	uint64_t file_end = file_segment->file_offset + file_segment->file_size;
+	char *p = buf;
 
-	if (offset > file_segment->file_size ||
-	    count > file_segment->file_size - offset) {
-		if (offset <= file_segment->file_size)
-			address += file_segment->file_size - offset;
+	if (!file_segment->zerofill && file_offset + count > file_end)
 		return drgn_error_create_fault("memory not saved in core dump",
 					       address);
-	}
-
-	uint64_t file_offset = file_segment->file_offset + offset;
-	char *p = buf;
 	while (count) {
-		ssize_t ret = pread(file_segment->fd, p, count, file_offset);
+		int readlen = min(file_end - file_offset, count);
```

Comment:
In theory, this can be over `INT_MAX`, so this should be `size_t`:

```suggestion
		size_t readlen = min(file_end - file_offset, count);
```